### ISSUE TITLE
Composition of contravariant functors

### ISSUE THREAD

**Issue by oysteins** [homalg-project/CAP_project#215]
I have three categories A, B and C, and two functors F : A -> B and G : B -> C, where F is covariant and G is contravariant.

I want to compose the two functors by calling `PreCompose( F, G )`.  This does not work, because CAP considers the contravariant functor G to be a covariant functor with B^op as source, and therefore it thinks F and G are not composable.  (Or, to quote precisely what CAP tells me, it "cannot decide whether morphisms are composable".)

Would it be possible to make CAP more aware of the functor's contravariantness, so that this functor could be composed with other functors in the expected way?

**Comment by sebastianpos** [comment]
Thanks for your suggestion. I will write up some remarks on this issue:

1) We should definitely provide a function that takes as an input a functor
F: A -> B
and outputs the functor
F^op: A^op -> B^op.

2) Such a function (call it `OppositeFunctor` for the moment) could then be used to
compose F: A -> B with G: B^op -> C by calling `PreCompose( OppositeFunctor( F ), G )`.

3) There could be a convenient function that checks whether F and G are composable 
in a covariant or contravariant way, and if so, does the composition. The creation of such a function is
possible, CAP is aware of the signature of its functors (see the examples added in #211)

I would refrain from calling such a convenient function `PreCompose`.
`PreCompose` is a CAP-operation and means the composition of morphisms in a CAP-category, and F:A->B and G:B^op -> C are simply not composable regarded as morphisms in the category of categories (as indicated by the inbuilt type check that raised the error method that you cited).
I would rather like to use a different name like `PreComposeFunctors` that is more specific for this context.

**Comment by oysteins** [comment]
I agree that `PreCompose` can not do what I want, since its behaviour should be consistent.  I'm wondering whether the real cause of the problem is that CAP forces me to think of functors as morphisms, which I don't really want to do.  I think it would be helpful (both in the situation described here and in other situations) if functors were not implemented as morphisms.  Since this is only tangentially related to the current issue, I'm writing that suggestion as a separate issue: #226
